src/linear_algebra/trace.lean

```diff
@@ -166,7 +166,39 @@ begin
   simp,
 end
 
-variables (M)
+theorem trace_prod_map :
+  trace R (M × N) ∘ₗ prod_map_linear R M N M N R =
+  (coprod id id : R × R →ₗ[R] R) ∘ₗ prod_map (trace R M) (trace R N) :=
+begin
+  let e := ((dual_tensor_hom_equiv R M M).prod (dual_tensor_hom_equiv R N N)),
+  have h : function.surjective e.to_linear_map := e.surjective,
+  refine (cancel_right h).1 _,
+  ext,
+  { simp only [dual_tensor_hom_equiv, tensor_product.algebra_tensor_module.curry_apply,
+  to_fun_eq_coe, tensor_product.curry_apply, coe_restrict_scalars_eq_coe, coe_comp,
+  linear_equiv.coe_to_linear_map, coe_inl, function.comp_app, linear_equiv.prod_apply,
+  dual_tensor_hom_equiv_of_basis_apply, map_zero, prod_map_apply, coprod_apply, id_coe, id.def,
+  add_zero, prod_map_linear_apply, dual_tensor_hom_prod_map_zero, trace_eq_contract_apply,
+  contract_left_apply, fst_apply] },
+  { simp only [dual_tensor_hom_equiv, tensor_product.algebra_tensor_module.curry_apply,
+  to_fun_eq_coe, tensor_product.curry_apply, coe_restrict_scalars_eq_coe, coe_comp,
+  linear_equiv.coe_to_linear_map, coe_inr, function.comp_app, linear_equiv.prod_apply,
+  dual_tensor_hom_equiv_of_basis_apply, map_zero, prod_map_apply, coprod_apply, id_coe, id.def,
+  zero_add, prod_map_linear_apply, zero_prod_map_dual_tensor_hom, trace_eq_contract_apply,
+  contract_left_apply, snd_apply], },
```

Comment:
If you added `@[simp]` to `dual_tensor_hom_prod_map_zero` and `zero_prod_map_dual_tensor_hom` (seems reasonable?), then this could just be `ext; simp`.

I wouldn't insist on this, however.

(I think unnecessarily squeezing simps can obfuscate proofs: it's hard to tell by looking at a big `simp only` if it is only there to speed things up, or there is real work happening.)

Reply:
Sounds fair. I was unsure for these kinds of proofs whether to prioritize conciseness with `simp` or speed with `simp only`. Since the difference in loading time was significant I chose `simp only` but I can totally change to simp if you believe that's a better practice.

Reply:
Sadness. Lean 3 is running out of steam.